adding support for index_select on quantized tensors

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61406**

Summary:
Only really needed to fix a few select functions so that they could work
for quantized tensors. Primarily creation and resizing of tensors
required a branch for quantized tensors. This doesn't work for
per_channel tensors

Test Plan: python test/test_quantization.py
TestQuantizedTensor.test_qtensor_index_select_cuda

python test/test_quantization.py
TestQuantizedTensor.test_qteensor_index_select_cpu
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29654446](https://our.internmc.facebook.com/intern/diff/D29654446)

HDCharles committed:
scoping for index_select on quantized tensors (WIP)

Summary: per_tensor was fairly straightforward for cuda, this is a really ugly hack to
show feasibility (it breaks the non quantized version though thats easy to fix). I don't forsee much
additional complication for cpu either though I didn't actually do that
test. per_channel is another story as you'd somehow need to apply
similar changes to the scale and zero_point tensors, unsure if thats in
scope or not. It'd probably be significantly more work.

Test Plan: python test/test_quantization.py
TestQuantizedtensor.test_qtensor_index_select

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "scoping for index_select on quantized tensors (WIP)"


Summary: per_tensor was fairly straightforward for cuda, this is a really ugly hack to
show feasibility (it breaks the non quantized version though thats easy to fix). I don't forsee much
additional complication for cpu either though I didn't actually do that
test. per_channel is another story as you'd somehow need to apply
similar changes to the scale and zero_point tensors, unsure if thats in
scope or not. It'd probably be significantly more work though.

Test Plan: python test/test_quantization.py
TestQuantizedtensor.test_qtensor_index_select

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29654446).

HDCharles commented:
This is intended as a fix for https://github.com/pytorch/pytorch/issues/59928

HDCharles committed:
Update on "adding support for index_select on quantized tensors"


Summary:
Only really needed to fix a few select functions so that they could work
for quantized tensors. Primarily creation and resizing of tensors
required a branch for quantized tensors. This doesn't work for
per_channel tensors

Test Plan: python test/test_quantization.py
TestQuantizedTensor.test_qtensor_index_select_cuda

python test/test_quantization.py
TestQuantizedTensor.test_qteensor_index_select_cpu
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29654446](https://our.internmc.facebook.com/intern/diff/D29654446)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29654446).

HDCharles committed:
Update on "adding support for index_select on quantized tensors"


Summary:
Only really needed to fix a few select functions so that they could work
for quantized tensors. Primarily creation and resizing of tensors
required a branch for quantized tensors. This doesn't work for
per_channel tensors

Test Plan: python test/test_quantization.py
TestQuantizedTensor.test_qtensor_index_select_cuda

python test/test_quantization.py
TestQuantizedTensor.test_qteensor_index_select_cpu
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29654446](https://our.internmc.facebook.com/intern/diff/D29654446)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -783,6 +783,32 @@ def test_qtensor_fill(self):
+            x = torch.randn([3] * dims, device=device) * 10
+            q = torch.quantize_per_tensor(x, scale, zp, quant_type)
+            qs = torch.index_select(q, index, selected)
+            for altered_index in range(2):
```
should we just compare the following two path matches:
1. float tensor -> index_select -> quantize
2. float tensor -> quantize -> index_select

HDCharles replied:
oh yeah thats much better, fixed in newest version

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29654446).

HDCharles committed:
Update on "adding support for index_select on quantized tensors"


Summary:
Only really needed to fix a few select functions so that they could work
for quantized tensors. Primarily creation and resizing of tensors
required a branch for quantized tensors. This doesn't work for
per_channel tensors

Test Plan: python test/test_quantization.py
TestQuantizedTensor.test_qtensor_index_select_cuda

python test/test_quantization.py
TestQuantizedTensor.test_qteensor_index_select_cpu
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29654446](https://our.internmc.facebook.com/intern/diff/D29654446)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29654446).

HDCharles committed:
Update on "adding support for index_select on quantized tensors"


Summary:
Only really needed to fix a few select functions so that they could work
for quantized tensors. Primarily creation and resizing of tensors
required a branch for quantized tensors. This doesn't work for
per_channel tensors

Test Plan: python test/test_quantization.py
TestQuantizedTensor.test_qtensor_index_select_cuda

python test/test_quantization.py
TestQuantizedTensor.test_qteensor_index_select_cpu
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29654446](https://our.internmc.facebook.com/intern/diff/D29654446)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29654446).

jerryzh168 reviewed (approved):
LG, thanks!
